examples/tabular_q.py

```diff
@@ -0,0 +1,169 @@
+    "A list of action names to explore from.",
+)
+flags.DEFINE_float("discount", 1.0, "The discount factor.")
+flags.DEFINE_list(
+    "features_indices",
+    [19, 22, 51],
+    "Indices of Alphaphase features that are used to construct a state",
+)
+flags.DEFINE_float("learning_rate", 0.1, "learning rate of the q-learning.")
+flags.DEFINE_integer("episodes", 5000, "number of episodes used to learn.")
+flags.DEFINE_float("epsilon", 0.2, "Epsilon rate of exploration. ")
+flags.DEFINE_integer("episode_length", 5, "The number of steps in each episode.")
+FLAGS = flags.FLAGS
+
+
+def hash_state_action(autophase_feature, action, step):
+    return tuple(
+        (
+            tuple(autophase_feature[FLAGS.features_indices]),
+            step,
+            FLAGS.actions.index(action),
+        )
+    )
+
+
+def get_env():
+    return env_from_flags(benchmark_from_flags())
+
+
+def select_action(q_table, ob, step, epsilon=0.0):
+    qs = [q_table.get(hash_state_action(ob, act, step), -1) for act in FLAGS.actions]
+    if random.random() < epsilon:
+        return random.choice(FLAGS.actions)
+    max_indices = [i for i, x in enumerate(qs) if x == max(qs)]
+    # Breaking ties at random by selecting any of the indices.
+    return FLAGS.actions[random.choice(max_indices)]
+
+
+def get_max_q_value(q_table, ob, step):
+    max_q = 0
+    for act in FLAGS.actions:
+        hashed = hash_state_action(ob, act, step)
+        if hashed not in q_table:
+            q_table[hashed] = 0
+        max_q = max(q_table[hashed], max_q)
+    return max_q
+
+
+def train(q_table, env):
+    # Buffer an old version of q table to inspect training progress
+    prev_q = {}
+
+    for i in range(FLAGS.episodes):
+        current_length = 0
+        obs = env.reset()
+        while current_length < FLAGS.episode_length:
+            # Run Epsilon greedy policy.
+            a = select_action(q_table, obs, current_length, FLAGS.epsilon)
+            hashed = hash_state_action(obs, a, current_length)
+            if hashed not in q_table:
+                q_table[hashed] = 0
+            obs, reward, done, info = env.step(env.action_space.flags.index(a))
+            current_length += 1
+            # Get max q at new state.
+            target = reward + FLAGS.discount * get_max_q_value(
+                q_table, obs, current_length
+            )
+            # Update Q value at current state action pair.
+            q_table[hashed] = (
+                FLAGS.learning_rate * target
+                + (1 - FLAGS.learning_rate) * q_table[hashed]
+            )
+
+        if i % 50 == 0:
+            print(f"Running episode {i}")
+
+            def compare_qs(q_old, q_new):
+                diff = [q_new[k] - v for k, v in q_old.items()]
+                return sum(diff) / len(diff) if diff else "NaN"
+
+            difference = compare_qs(prev_q, q_table)
+            print(
+                f"Newly added Q entries {len(q_table)-len(prev_q)}, averaged diff {difference}"
```

Comment:
As well as logging the progression of the Q-table size/values, would it make sense to log some kind of measure of the quality of the trajectories evaluated? I had a little play around with buffering a batch of rewards and logging the average, maybe something like this would be nice:

```
episode=  50, avg_reward=0.6836, Q-table_entries= 1131, Q-table_diff=0.0015929
episode= 100, avg_reward=0.8026, Q-table_entries= 1945, Q-table_diff=0.0001957
episode= 150, avg_reward=0.8455, Q-table_entries= 2428, Q-table_diff=0.0002611
episode= 200, avg_reward=0.8591, Q-table_entries= 2803, Q-table_diff=0.0002467
...
episode=4950, avg_reward=0.9705, Q-table_entries=10365, Q-table_diff=0.0000241
episode=5000, avg_reward=0.9422, Q-table_entries=10380, Q-table_diff=0.0000213
```

done using:

```py
    prev_q = {}
    rewards = []
    for i in range(1, FLAGS.episodes + 1):
        observation = env.reset()
        for step in range(FLAGS.episode_length):
            # ...

        rewards.append(env.episode_reward)
        if i % FLAGS.log_every == 0:
            # ...
            avg_reward = sum(rewards) / len(rewards)
            rewards = []
            print(
                f"episode={i:4d}, avg_reward={avg_reward:.4f}, "
                f"Q-table_entries={len(q_table):5d}, Q-table_diff={difference:.7f}"
            )
```

I'm not sure if reward average makes sense, and either way it isn't a big deal at all

Reply:
Typically in RL one would *evaluate* the policy by rolling out in the environment and observe its performance. I'm adding a rollout function and reporting the resulting rewards of the trajectory. It's a good point though.

Reply:
Sounds good, thanks!